Add a new type of switching function with an adjustable weight parameter

yipenggao opened the pull request:
Add a new type of switching function (MIX) with an adjustable weight parameter (mix_weight).

Add a new inputfile (modified from DerivativeMultiPhaseMaterial.i) to test the new function.

closes #9042

yipenggao commented:
This my first time to change the source code file. Please let me know if it is not within the code standard. :)

Thanks.

yipenggao commented:
Ok. It fails all the checks.... I will work on it next weekend... :(

dschwen commented:
You need to activate the new test by putting it into the tests file. I have more comments later.

yipenggao commented:
@dschwen
Should I put create a new folder under /tests/ and put the input and output file there? Also a file named "tests" is needed, right?

dschwen commented:
No new directory. Look where the other modes of the switching function are tested and make your test input similar to those.

dschwen commented on modules/phase_field/src/materials/SwitchingFunctionMaterial.C:
```diff
@@ -11,18 +11,24 @@ InputParameters
       "h_order", h_order, "Polynomial order of the switching function h(eta)");
   params.set<std::string>("function_name") = std::string("h");
+
+  Real mix_weight = 1;
```
remove this variable

dschwen commented on modules/phase_field/src/materials/SwitchingFunctionMaterial.C:
```diff
@@ -46,6 +52,14 @@ SwitchingFunctionMaterial::computeQpProperties()
       break;
 
+    case 2: // MIX
+      _mix_weight = _mix_weight > 1 ? 1 : (_mix_weight < 0 ? 0 : _mix_weight);
```
No. Range limit the parameter, set it in the initializer list, and naver change it again. In fact make it `const` in the declaration.

dschwen commented:
No new directory. Look where the other modes of the switching function are tested and make your test input similar to those.

dschwen commented:
I'd actually rather see this implemented in a derived class. The mix weight parameter is only used for the one new h order option.

dschwen commented:
Why did you close this?

yipenggao commented:
@dschwen 
... Sorry, I reset it to previous version (then push....), and right now I am working on adding a new class files. I didn't realize that the PR closed automatically...

Can we reopen it or we have to make a new PR?

yipenggao committed:
Add a new class SwitchingFunctionMaterialMix to calculate a few mix types of switching function.
Tests files are also included.

Add output file to /gold/ folder which closes #9042.

Code style corrected which closes #9042.

Correct typos in tests file.

Reduce the total steps of running to limit numerical errors.

Shorten total number of evolution steps to limit numerical error.

Change weight parameter to shorten the convergence process.

Update MixedSwitchingFunctionMaterial class files.
Add test inputfile mixedswitchingfunctionmaterial.i and corresponding output file in /gold/.
Update tests file.
Update code format.

yipenggao commented:
It still fails the check... Any hint about why?

permcody commented:
Click on the precheck target details. It'll take you to CIVET where you can open the results to look at them. You just need a ticket reference in one or more of your commits in this PR.

permcody commented:
Also we prefer "rebase" strategy with MOOSE development, not "merge" strategy.

moosebuild commented:
Job `Precheck:linux-gnu` on 3cb3dc6 wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 40a9dac5d3df30c8b3278c9b8163aebb2cf9128c`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/9116/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/9116/style.patch | git apply -v`

moosebuild commented:
Job `Moose Documentation:linux-clang` on 069af20 wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9116/site

yipenggao commented:
@permcody 
Thanks for your help!

Most of the fail information are "Missing gold file".

I have put the inputfile in .../phase_field/tests/MultiPhase/ , and the output file in .../phase_field/tests/MultiPhase/gold/

Is that the correct way to do it?

permcody commented on modules/phase_field/tests/MultiPhase/tests:
```diff
@@ -65,4 +65,9 @@
+  [./DerivativeMultiPhaseMaterial_SwitchingMix]
+    type = 'Exodiff'
+    input = 'DerivativeMultiPhaseMaterial_SwitchingMix.i'
+    exodiff = 'DerivativeMultiPhaseMaterial_SwithcingMix_out.e'
```
The problem is spelling 'Swithcing' -> 'Switching'

yipenggao commented:
@permcody 

Many thanks for your review. I will be more careful.

moosebuild commented:
Job `Moose Documentation:linux-clang` on f56162e wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9116/site

yipenggao commented:
It is said "Failed(exodiff)".

Is that because I use incorrect type option in the tests file?
type = 'Exodiff'

permcody commented:
No, it's because it actually is diffing. Scroll up in the output and look at the differences. They are very tiny (in the e-10 range) which is right near our absolute zero value. The Exodiff tolerances can be adjusted but you generally shouldn't change them. The defaults are listed in the Exodiff object itself. There's a relative tolerance and an absolute floor.

Instead you might re-evaluate your test and decide whether or not you need that many steps to actually test your physics (over 30). The first 20 or so are identical (this is all in the output). You start to get a little numerical round-off as the small differences accumulate. There are many ways to make all of these things go away. If you have any questions about the Exodiff output don't hesitate to ask. It's not too difficult to look through though. 

What you are experiencing is common by the way. You are probably creating tests on a Mac but the test systems are mostly Linux.

moosebuild commented:
Job `Moose Documentation:linux-clang` on 876d95a wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9116/site

permcody commented:
@yipenggao - As you shorten up the test, you'll need to update the gold file too. Exodiff complains if the number of output steps don't match.

moosebuild commented:
Job `Moose Documentation:linux-clang` on 903a0c5 wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9116/site

dschwen commented on modules/phase_field/include/materials/SwitchingFunctionMaterialMix.h:
```diff
@@ -0,0 +1,40 @@
+ * \see KKSPhaseChemicalPotential
+ * \see KKSCHBulk
+ */
+class SwitchingFunctionMaterialMix : public OrderParameterFunctionMaterial
```
Please rename to `MixedSwitchingFunctionMaterial`; this should end on Material for consistency.

dschwen commented on modules/phase_field/include/materials/SwitchingFunctionMaterialMix.h:
```diff
@@ -0,0 +1,40 @@
+  MooseEnum _h_order;
+
+  /// Weight parameter of mixed-type h(eta)
+  Real _weight;
```
make `const`

dschwen commented on modules/phase_field/src/materials/SwitchingFunctionMaterialMix.C:
```diff
@@ -0,0 +1,54 @@
+                             "polynomial forms.\nMIX234: "
+                             "weight*(3*eta^2-2*eta^3)+(1-weight)(4*eta^3)-3*eta^4");
+  MooseEnum h_order("MIX234=0", "MIX234");
+  params.addParam<MooseEnum>(
```
remove the `h_order` parameter. No point in having a single option parameter.

dschwen commented on modules/phase_field/src/materials/SwitchingFunctionMaterialMix.C:
```diff
@@ -0,0 +1,54 @@
+  Real n = _eta[_qp];
+  n = n > 1 ? 1 : (n < 0 ? 0 : n);
+
+  switch (_h_order)
```
if `_h_order` is removed the switch block becomes unnecessary

dschwen commented on modules/phase_field/tests/MultiPhase/DerivativeMultiPhaseMaterial_SwitchingMix.i:
```diff
@@ -0,0 +1,385 @@
+[Mesh]
```
model you test based on `switchingfunction3phasematerial.i` for example. You do not need a solve to test this material. Please try to keep to the standard for filename capitalization and use of underscores for at least the directory the file is located in.

moosebuild commented:
Job `Moose Documentation:linux-clang` on a757943 wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9116/site

dschwen commented:
Oh, right, before this becomes mergeable it needs a documentation file! I can show you how to get going on that.

moosebuild commented:
Job `Moose Documentation:linux-clang` on b69768a wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9116/site

moosebuild commented:
Job `Moose Documentation:linux-clang` on b41601e wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9116/site

moosebuild commented:
Job `Precheck:linux-gnu` on 220ee25 wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 150c068c96bbd7a8d1b72e7dba57d492069362f1`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/9116/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/9116/style.patch | git apply -v`

dschwen commented:
When you apply the style fixes please _squash_ your commits as much as it makes sense.

moosebuild commented:
Job `Moose Documentation:linux-clang` on 412f863 wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9116/site

yipenggao commented:
Great! It finally passes all checks.

Then how to make documentation file. Any online instruction?

yipenggao committed:
Add documentation for MixedSwitchingFunctionMaterial with detailed descriptions.

moosebuild commented:
Job `Moose Documentation:linux-clang` on 5667ad1 wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9116/site

dschwen commented on docs/content/documentation/systems/Materials/phase_field/MixedSwitchingFunctionMaterial.md:
```diff
@@ -0,0 +1,17 @@
+$$
+k(2\eta^2-\eta^4)+(1-k)(3\eta^4-2\eta^6)
+$$
+
```
please at least mention MIX234 and MIX246 here. Just two terms is a bit weak for documentation.

moosebuild commented:
Job `Moose Documentation:linux-clang` on e10b37c wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9116/site
